### PR TITLE
feat(chart): add servicemonitor

### LINKED ISSUE #2313: Add ServiceMonitor to chart
**Is your feature request related to a problem?**
It'd be great if the Helm chart had a `ServiceMonitor` resource that could be enabled.

**Describe the solution you'd like**
Add an optional `ServiceMonitor` resource to the chart.

**Describe alternatives you've considered**
I currently have to manually create this resource.

### PR DESCRIPTION
**Pull request by stevehipwell**
### Issue

Fixes #2313.
Fixes [aws/eks-charts#498](https://github.com/aws/eks-charts/issues/498)

Replaces [aws/eks-charts#499](https://github.com/aws/eks-charts/pull/499)

### Description

This PR adds an optional `ServiceMonitor` and the service required to collect the metrics. As part of this work I've refactored the current webhook service into a new file with templated naming; this would allow this extra service to be removed in favour of the standard service in the future.

- Increment chart patch version (based on past PRs, personally this would be a minor update for me)
- Add optional `ServiceMonitor` & optional (for now) standard `Service`
- Add new `aws-load-balancer-controller.webhookService` template to move this logic into one place
- Move webhook `Service` to new file to make it's purpose clearer
- Add extra labels to webhook service to distinguish it from the standard `Service`

### Checklist
- [ ] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [x] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by stevehipwell** [comment]
@kishorj @M00nF1sh could you take a look at this?

**Comment by stevehipwell** [comment]
@kishorj @M00nF1sh I'm still waiting on some feedback for this?

**Comment by stevehipwell** [comment]
@kishorj @M00nF1sh I've rebased this, could we look at getting it merged?

**Comment by kishorj** [comment]
> @kishorj @M00nF1sh I've rebased this, could we look at getting it merged?

I was busy with the v2.3.1 release. Now it is out, I will look into it this week.

**Comment by stevehipwell** [comment]
@kishorj have you had a chance to look over this?

File: helm/aws-load-balancer-controller/templates/service-webhook.yaml
Comment (stevehipwell) [reply helm/aws-load-balancer-controller/templates/service-webhook.yaml:1]: > why not combine this in the service.yaml?

I covered this in the description. The current service has been named for the webhook even though it's just a service for the deployment. This PR doesn't change the webhook being in it's own isolated service, but I'm planning on opening a PR once it's merged to unify the services in _service.yaml_. Would it be easier if I did this directly in this PR?

> If we need a new file, webhook-service.yaml would sound more appropriate

The standard idiom is for files to have a resource type prefix (or just be named for the resource). If you have separate components they are idiomatically separated into sub folders. In this case the webhook isn't a separate component and should really be in _service.yaml_ (see above).

File: helm/aws-load-balancer-controller/templates/service.yaml
Comment (stevehipwell) [reply helm/aws-load-balancer-controller/templates/service.yaml:5]: Services shouldn't be used like this, this is the service for the deployment and should have all of the ports in the one service object. See above.

File: helm/aws-load-balancer-controller/templates/servicemonitor.yaml
Comment (stevehipwell) [reply helm/aws-load-balancer-controller/templates/servicemonitor.yaml:2]: I agree about the documentation, but testing the API server often doesn't work as expected (`helm template` and ArgoCD come to mind) so I'd advise against this approach.

Comment (kishorj) [reply helm/aws-load-balancer-controller/templates/servicemonitor.yaml:2]: It may not work with ArgoCD, but IMO still valuable for helm install/upgrade workflows. The `helm template` with `--validate` flag also honors the api server capabilities. For workflows that doesn't support API capabilities check, doesn't work either way, so why not include the check?

File: helm/aws-load-balancer-controller/templates/service-webhook.yaml
Comment (kishorj) [reply helm/aws-load-balancer-controller/templates/service-webhook.yaml:1]: if we need to rename the webhook service, then lets not combine the services together for now. I'd prefer to either have both the service manifests in `servicve.yaml` or include the new service spec in the `servicemonitor.yaml` file.

File: helm/aws-load-balancer-controller/templates/servicemonitor.yaml
Comment (stevehipwell) [reply helm/aws-load-balancer-controller/templates/servicemonitor.yaml:2]: @kishorj every time I've added complex checks into charts the ArgoCD users (or users who don't know to use `helm template --validate`) have pushed back. I'm happy to add the check as it'd work for my use case, but I just want to make sure that we've considered this as a downside. Personally I'd still leave it without the check as it's an opt-in setting that will either work or error so the cost seems to outweigh the benefit but it's your call.

File: helm/aws-load-balancer-controller/templates/_helpers.tpl
Comment (stevehipwell) [reply helm/aws-load-balancer-controller/templates/_helpers.tpl]: Fixed.

File: helm/aws-load-balancer-controller/templates/service-webhook.yaml
Comment (stevehipwell) [reply helm/aws-load-balancer-controller/templates/service-webhook.yaml]: I've combined both in _service.yaml_ and will open a separate PR after this is merged to remove the webhook service as it's a duplicate.

File: helm/aws-load-balancer-controller/templates/service.yaml
Comment (kishorj) [inline helm/aws-load-balancer-controller/templates/service.yaml:14]: cert won't be valid for this service, so opening this port does not serve any purpose.

File: helm/aws-load-balancer-controller/templates/service.yaml
Comment (kishorj) [reply helm/aws-load-balancer-controller/templates/service.yaml:4]: The new service should be configured only if service monitor is enabled.

File: helm/aws-load-balancer-controller/templates/servicemonitor.yaml
Comment (kishorj) [reply helm/aws-load-balancer-controller/templates/servicemonitor.yaml:2]: if the CRD is not installed, but user enables serviceMonitor, does the chart installation fail or we'd end up with a partial installation?

**Review by kishorj — COMMENTED** [review]
Please include a section in the README with the steps to install the monitoring.coreos.com/v1 CRD.

File: helm/aws-load-balancer-controller/templates/service.yaml
Comment (stevehipwell) [reply helm/aws-load-balancer-controller/templates/service.yaml:14]: Good point, I'll remove this for now.

File: helm/aws-load-balancer-controller/templates/servicemonitor.yaml
Comment (stevehipwell) [reply helm/aws-load-balancer-controller/templates/servicemonitor.yaml:2]: I will check, but the behaviour as coded would be consistent with the other use of CRDs for cert-manager which only checks for v1 not if the CRD type exists at all.

**Comment by stevehipwell** [comment]
@kishorj I've made the changes you requested if you could re-review.

**Comment by stevehipwell** [comment]
@kishorj sorry to comment again but could you please take a look a this as it should be ready to merge?

**Comment by kishorj** [comment]
> @kishorj sorry to comment again but could you please take a look a this as it should be ready to merge?

@stevehipwell, I'm waiting for two changes from your side, not sure if you have them locally and did not push to the PR yet.
1. remove the port 443 from the new service since the certs won't be valid
2. A new section in the README with the steps to install the monitoring.coreos.com/v1 CRD.

**Comment by kishorj** [comment]
@stevehipwell, thanks for the changes. Regarding the cert-manager, we can update the README separately on a different PR. The installation guide mentions the following for yaml manifest installation option:
```
Install cert-manager

kubectl apply --validate=false -f https://github.com/jetstack/cert-manager/releases/download/v1.5.3/cert-manager.yaml
```
We can mention similar steps for helm installation as well.

Since ServiceMnoitor is a new feature for this chart, I feel it would be prudent to include a section with the necessary installation steps if users set the serviceMonitor.enabled=true.

**Comment by stevehipwell** [comment]
@kishorj I've updated the chart README to cover both cert-manager and the Prometheus Operator ServiceMonitor CRD from the context of a Helm user.

**Review by kishorj — APPROVED** [review]

**Pull request merged**
